src/tests/architect_tests/test_cohort_table_generators.py

```diff
@@ -98,3 +99,97 @@ def test_cohort_table_generator():
         assert results == expected_output
         utils.assert_index(engine, table_generator.cohort_table_name, "entity_id")
         utils.assert_index(engine, table_generator.cohort_table_name, "as_of_date")
+
+        table_generator.generate_cohort_table(as_of_dates)
+        assert results == expected_output
+
+
+def test_cohort_table_generator_noreplace():
+    input_data = [
+        (1, datetime(2016, 1, 1), True),
+        (1, datetime(2016, 4, 1), False),
+        (1, datetime(2016, 3, 1), True),
+        (2, datetime(2016, 1, 1), False),
+        (2, datetime(2016, 1, 1), True),
+        (3, datetime(2016, 1, 1), True),
+        (5, datetime(2016, 3, 1), True),
+        (5, datetime(2016, 4, 1), True),
+    ]
+    with testing.postgresql.Postgresql() as postgresql:
+        engine = create_engine(postgresql.url())
+        utils.create_binary_outcome_events(engine, "events", input_data)
+        table_generator = CohortTableGenerator(
+            query="select entity_id from events where outcome_date < '{as_of_date}'::date",
+            db_engine=engine,
+            cohort_table_name="exp_hash_cohort",
+            replace=False
+        )
+
+        # 1. generate a cohort for a subset of as-of-dates
+        as_of_dates = [
+            datetime(2016, 1, 1),
+            datetime(2016, 2, 1),
+            datetime(2016, 3, 1),
+        ]
+        table_generator.generate_cohort_table(as_of_dates)
+        expected_output = [
+            (1, datetime(2016, 2, 1), True),
+            (1, datetime(2016, 3, 1), True),
+            (2, datetime(2016, 2, 1), True),
+            (2, datetime(2016, 3, 1), True),
+            (3, datetime(2016, 2, 1), True),
+            (3, datetime(2016, 3, 1), True),
+        ]
+        results = list(
+            engine.execute(
+                f"""
+                select entity_id, as_of_date, active from {table_generator.cohort_table_name}
+                order by entity_id, as_of_date
+            """
+            )
+        )
+        assert results == expected_output
+        utils.assert_index(engine, table_generator.cohort_table_name, "entity_id")
+        utils.assert_index(engine, table_generator.cohort_table_name, "as_of_date")
+
+        table_generator.generate_cohort_table(as_of_dates)
+        assert results == expected_output
+
+        # 2. generate a cohort for a different subset of as-of-dates,
```

Comment:
If you change your temporal config, you will use a different cohort table. So the only time this should happen is if your underlying data change in such a way that there were some as_of_dates that previously had no entities in the cohort and now they do. I mean, no objection to testing for it as it's clearly the intended behavior. But this should, really, never happen.

Reply:
Sure, this is pretty unlikely as it stands now. The most likely way this behavior would show up, I think, is if your experiment died (or you killed it) during cohort table building. I do get that cohort table building is not a common place for experiments to die, so this is pretty much an edge case.

In making this test I was also thinking about what happens when we do switch the cohort table to be namespaced based on the cohort query and/or name and not the entire experiment. Since this component expects to be passed in a table name it is unaware of the external logic used to determine the table name, and should work regardless.

Reply:
Ah, good points. I had forgotten about failed experiments somehow